### ISSUE TITLE
2009 creation dates returned when they shouldn't be

### ISSUE THREAD

**Issue by unpredictable** [Electroid/mojang-api#34]
According to [jomo's creation date finder](https://gist.github.com/jomo/be7dbb5228187edbb993) which I believe was implemented into this API (or something similar was implemented), January 10th, 2010 is the earliest date where accounts can be created. The user "Mixo" (https://api.ashcon.app/mojang/v2/user/Mixo) returns a creation date in 2009 when it shouldn't be returning one. Plus, the user is legacy, and legacy users can't have their creation dates tracked via the API. For most legacy accounts this API returns `null` as the `created_at` value, but for whatever reason this isn't the case for Mixo.

Also, using this API you should be able to see whether a user is legacy/unmigrated or not, this feature should be added soon especially considering the MS migration.

The API also returns incorrect naming histories for accounts I have noticed, for example the user `0d565263382f433687eb45c9c6375eff` with a name history of 23 names only had their current name in the history according to this API. (`[{'username': 'myeve____'}]`) This issue has since fixed itself, and I presume it to be a ratelimit issue, but if this API is ratelimited from any Mojang endpoints it should return 429 to avoid returning incorrect API responses, at least in my opinion.

**Comment by Electroid** [comment]
Thanks for flagging this, yes the caching will need to be adjusted. I do plan on issuing an update soon, especially with the Microsoft migration around the corner.

**Comment by unpredictable** [comment]
Any ETA on when? Hoping soon considering I have a lot of UUIDs I’d like to pass through to the API

**Comment by unpredictable** [comment]
Quick update to this case.

The method that is used to fetch creation date has stopped working (see https://bugs.mojang.com/browse/WEB-3367 for more information).

It was stated that the caching will need to be adjusted. Is it possible for you to avoid adjusting the cache for most accounts, as they are no longer able to be re-grabbed from the API, only removing creation dates for accounts "created" in 2009? Those creation dates *should* be null but for some e.g Jog then there *was* a way to fetch the dates for these (we call them giftcoded accounts/giftcode snipes) - if you'd like to see what we used to do (again sadly not possible anymore), you can check 88/mcreation.

Also, just out of curiosity - how many accounts are in the API's cache currently?

**Comment by Electroid** [comment]
Thanks for linking that issue. That's concerning, I hope they fix it!

I might punt a `v3` because I'm assuming Microsoft/Mojang may change the API. Although, I am considering just adding the `legacy` tag for `v2` in the mean time.

There are millions of "birthdays", I can look at invalidating the impossible dates this weekend.

**Comment by unpredictable** [comment]
Hoping they fix it too. 

Please do add the legacy/demo tag for v2 while we wait for the new API. Would be very useful, and I don’t think it’s too hard? Hit /user/profile/<uuid> I think... and that’s it. You can use the session server endpoint with just legacy but the demo tag is helpful as well.

The impossible dates are anything in 2009. 2010 January 10 isn’t impossible just to clarify.

Also how many accounts do you have cached currently?

**Comment by Electroid** [comment]
I just deployed `legacy` and `demo` fields. If either of those is true, `created_at` will be `null`. So I think that solves most of these edge cases.

Going to keep this open until Mojang fixes the other issue.

**Comment by unpredictable** [comment]
Any chance we can get invalid username support as well? I have users like Mr.LoL, crazybitingturtle, and glenn-stroek in my database and would like to see records for them via the API. Should be an easy regex change from just looking at where username validation takes place in this code?

**Comment by unpredictable** [comment]
Reporting yet another issue. A user of mine reported to me that their birthday was (very clearly) inaccurate. Don't mind the duplicate name history towards the end. That's a glitch with Mojang removing a name from their history and that does show up in the actual API. 

Their API URL is: https://api.ashcon.app/mojang/v2/user/lifeispain_326

You can clearly see the birthday is inaccurate considering that the birthday is August 8, 2015... and they changed their name 7 times before that date.

**Comment by unpredictable** [comment]
Yeah, something's surely up. "Nimble" is reported to have been made on April 16, 2020, when it was name-changed in 2015.

**Comment by Electroid** [comment]
Alright, thanks for flagging again. My hunch is that sporadic `429`s might be "corrupting" some birthdays. In which case, I'll have the purge the cache anyway. Will update when these is progress.

**Comment by unpredictable** [comment]
Yeah, the accounts with birthdays after name changes should have their caches purged, if you know what I mean?

Like, Nimble has a name change in 2015 but its birthday is 2020. If the birthday date surpasses the date of the first name change then the cache should be purged for that account.
